crates/runtime/src/memfd.rs

```diff
@@ -97,124 +68,107 @@ impl ModuleMemFds {
+        let result = module.memory_initialization.init_memory(
+            |memory| module.memory_plans[memory].memory.minimum,
+            // Right now memfd initialization isn't supported for initializers
+            // that have a global as the offset.
+            |_global| None,
+            |memory, offset, data_range| {
+                // Memfd-based initialization of an imported memory isn't
+                // implemented right now, although might perhaps be
+                // theoretically possible for statically-known-in-bounds
+                // segments with page-aligned portions.
+                let memory = match module.defined_memory_index(memory) {
+                    Some(index) => index,
+                    None => return InitMemory::NoImportedMemory,
+                };
+
+                // Splat the `data_range` into the `image` for this memory,
+                // updating it as necessary with 0s for holes and such.
+                let image = &mut images[memory];
+                let data = &wasm_data[data_range.start as usize..data_range.end as usize];
+                let offset = offset as usize;
+                let new_image_len = offset + data.len();
+                if image.len() < new_image_len {
+                    if new_image_len > MAX_MEMFD_IMAGE_SIZE {
+                        return InitMemory::Error;
                     }
-                    let memfd = memfds[defined_memory].as_mut().unwrap();
-
-                    let end = round_up_page(segment.end().expect("must have statically-known end"));
-                    if end > sizes[defined_memory] {
-                        sizes[defined_memory] = end;
-                        memfd.as_file().set_len(end)?;
-                    }
-
-                    let base = segments[i].offset;
-                    let data = &wasm_data[segment.data.start as usize..segment.data.end as usize];
-                    memfd.as_file().write_at(data, base)?;
+                    image.resize(new_image_len, 0);
                 }
-            }
-            &MemoryInitialization::Paged { ref map, .. } => {
-                for (defined_memory, pages) in map {
-                    let top = pages
-                        .iter()
-                        .map(|(base, range)| *base + range.len() as u64)
-                        .max()
-                        .unwrap_or(0);
-
-                    let memfd = create_memfd()?;
-                    memfd.as_file().set_len(top)?;
-
-                    for (base, range) in pages {
-                        let data = &wasm_data[range.start as usize..range.end as usize];
-                        memfd.as_file().write_at(data, *base)?;
-                    }
-
-                    memfds[defined_memory] = Some(memfd);
-                    sizes[defined_memory] = top;
-                }
-            }
+                image[offset..][..data.len()].copy_from_slice(data);
```

Comment:
interesting idiom here -- any reason not to do `[offset..(offset+data.len())]`? This is kind of elegant but it took me a second to parse (my brain wanted to see it as a 2D array somehow at first).

Reply:
I've done this in a few other places in Wasmtime as well but the main benefits in my opinion are:

* Slightly more ergonomic (don't have to repeat the base twice)
* Guaranteed no overflow (no additions which could accidentally overflow)

Although I agree that if it's the first time seeing this it can be a bit odd.

Reply:
I'll take this as a "learned a nice Rust idiom from Alex" moment then :-)